stats/cloud/collector.go

```diff
@@ -269,6 +269,7 @@ func (c *Collector) Collect(sampleContainers []stats.SampleContainer) {
 	for _, sampleContainer := range sampleContainers {
 		switch sc := sampleContainer.(type) {
 		case *httpext.Trail:
+			sc.Tags.ReplaceURLWithName()
```

Comment:
This is (IMO) *obviously* racy with any other output that decides to do anything with the tags ... I also tested it but the 
callstack is way too long (json/encoding takes forever): 

```
WARNING: DATA RACE
Read at 0x00c005d9ba30 by goroutine 33:
  encoding/json.compact()
      /home/mstoykov/.gvm/gos/go1.12.9/src/encoding/json/indent.go:20 +0x1c3
  encoding/json.marshalerEncoder()
      /home/mstoykov/.gvm/gos/go1.12.9/src/encoding/json/encode.go:457 +0x216
  encoding/json.structEncoder.encode()
# more lines inside json/encode.go
 encoding/json.Marshal()                                                                                                                                                                                [94/373771]
      /home/mstoykov/.gvm/gos/go1.12.9/src/encoding/json/encode.go:160 +0x73
  github.com/loadimpact/k6/stats/cloud.(*Client).PushMetric()
      /home/mstoykov/.gvm/pkgsets/go1.12.9/global/src/github.com/loadimpact/k6/stats/cloud/api.go:104 +0x905
  github.com/loadimpact/k6/stats/cloud.(*Collector).pushMetrics()
      /home/mstoykov/.gvm/pkgsets/go1.12.9/global/src/github.com/loadimpact/k6/stats/cloud/collector.go:498 +0x40b
  github.com/loadimpact/k6/stats/cloud.(*Collector).Run()
      /home/mstoykov/.gvm/pkgsets/go1.12.9/global/src/github.com/loadimpact/k6/stats/cloud/collector.go:248 +0x168
  github.com/loadimpact/k6/core.(*Engine).Run.func1()
      /home/mstoykov/.gvm/pkgsets/go1.12.9/global/src/github.com/loadimpact/k6/core/engine.go:149 +0x52


Previous write at 0x00c005d9ba30 by goroutine 31:
  runtime.slicecopy()
      /home/mstoykov/.gvm/gos/go1.12.9/src/runtime/slice.go:197 +0x0
  encoding/json.Marshal()
      /home/mstoykov/.gvm/gos/go1.12.9/src/encoding/json/encode.go:164 +0x135
  github.com/loadimpact/k6/stats.(*SampleTags).MarshalJSON()
      /home/mstoykov/.gvm/pkgsets/go1.12.9/global/src/github.com/loadimpact/k6/stats/stats.go:236 +0xb1
  encoding/json.marshalerEncoder()
      /home/mstoykov/.gvm/gos/go1.12.9/src/encoding/json/encode.go:454 +0xea
# again more lines in json/encode.go
      /home/mstoykov/.gvm/gos/go1.12.9/src/encoding/json/encode.go:306 +0xa6
  encoding/json.(*Encoder).Encode()
      /home/mstoykov/.gvm/gos/go1.12.9/src/encoding/json/stream.go:200 +0xd7
  github.com/loadimpact/k6/stats/json.(*Collector).commit()
      /home/mstoykov/.gvm/pkgsets/go1.12.9/global/src/github.com/loadimpact/k6/stats/json/collector.go:154 +0x394
  github.com/loadimpact/k6/stats/json.(*Collector).Run()
      /home/mstoykov/.gvm/pkgsets/go1.12.9/global/src/github.com/loadimpact/k6/stats/json/collector.go:109 +0xf3
  github.com/loadimpact/k6/core.(*Engine).Run.func1()
      /home/mstoykov/.gvm/pkgsets/go1.12.9/global/src/github.com/loadimpact/k6/core/engine.go:149 +0x52

Reply:
But considering that the engine calls the `Collect()` method sequentially: https://github.com/loadimpact/k6/blob/9e1c5c11149c699fbbe601d8a724f7a76a869d09/core/engine.go#L393-L395

... and `sc.Tags.ReplaceURLWithName()` is called in `Collect()`, I don't think we need to be worried about race conditions.

Reply:
@na-- the above is the actual calltrace and you are changing the tags of the sample that is *shared* between all the outputs so all ouputs get a `url` that is equal to `name`. I would say that this shouldn't happen either way. 
More importantly Collect is called sequentially but practically all outputs buffer the samples separately and write them in another goroutine concurrently with one another

Reply:
@MStoykov Can you post code to reproduce the race condition? It relates to the changes in this CL, or present already?

Reply:
I installed your code with `-race` and ran `k6 run script.js -o json=tr.json.gz -o cloud`

Reply:
and now the json outputs has all the `url`s equal to `name` which it didn't if I remove the `-o cloud`

Reply:
@MStoykov but the stacktrace does not point to any changes in this PR. Or I miss something?

Reply:
hmm yeah, seems like the JSON is actually the issue here :disappointed_relieved:

Reply:
Yeah ... currently the part that we just write to a field is the racy part that is caught in the race stacktrace I copied ... dunno if the other is caught as it's basically another 50 lines + the amount of data races covers my whole screen. 
I am gonna try and do it without touching the original tags and possibly add a performance penalty and check how big of a performance penalty it's adding

Reply:
The problem seems not due to the changing of map, but the invalidate json cache, at `st.json = nil`.

Reply:
That could theoretically cause a problem, if the cloud's `Collect()` execution coincides with the pushing of metrics from a previous collector, but the specific stacktrace @MStoykov posted happens just because of the way we cache the JSON without synchronization, i.e. it could happen on the current `master` as well...

Reply:
@na-- if it's true, then we should merge this and create new issue to track and fix it?

Reply:
Not sure... given that this slightly complicates the mess, and it this is potentially a panic even when k6 is not compiled with `-race`, maybe we should fix both now? Or at least implement something like @MStoykov's first suggestion below, so that this PR doesn't make the situation worse. While fixing the json caching issue would be trivial with a `sync.RWMutex`, it probably needs to be benchmarked. And the fix in this PR probably also needs to be benchmarked...

Reply:
I confirm that the problem exist on master. I don't think benchmark is necessary, we have to make it works correctly first.

Reply:
Of course correct work is more important. By benchmarking I meant comparing the different fixes :wink: For example, would it make more sense to continue caching the JSON result, but in a thread-safe way, or should we just ditch that logic, etc.

Reply:
@na-- FYI, here's the benchmark

**Remove the cache**

```
name                       old time/op    new time/op       delta
SampleTags_MarshalJSON-12    2.38ns ± 4%  415169.75ns ± 2%  +17414738.51%  (p=0.000 n=10+8)

name                       old alloc/op   new alloc/op      delta
SampleTags_MarshalJSON-12     0.00B        112151.00B ± 0%          +Inf%  (p=0.000 n=10+10)

name                       old allocs/op  new allocs/op     delta
SampleTags_MarshalJSON-12      0.00           2007.00 ± 0%          +Inf%  (p=0.000 n=10+10)
```

**Using sync.Mutex**

```
name                       old time/op    new time/op    delta
SampleTags_MarshalJSON-12    2.38ns ± 4%   12.36ns ± 5%  +418.46%  (p=0.000 n=10+10)

name                       old alloc/op   new alloc/op   delta
SampleTags_MarshalJSON-12     0.00B          0.00B           ~     (all equal)

name                       old allocs/op  new allocs/op  delta
SampleTags_MarshalJSON-12      0.00           0.00           ~     (all equal)
```

Seems `sync.Mutex` is a better choice.

Reply:
`sync.RWMutex` would probably be even better than a `sync.Mutex`. And in either case, this doesn't actually answer the question how often the caching is actually used in a real-life k6 execution. At least a few, since for the non-cloud outputs, the same tags of an HTTP metric sample would be marshaled multiple ~8 times, once for [each `htto_req_*` distinct metric](https://github.com/loadimpact/k6/blob/78858ef41d2927f43db4f10ca64f791cf7f329ce/lib/metrics/metrics.go#L42-L49).

 Anyway, as you said, fixing the JSON data race issue could be a separate PR. For this specific PR, I'm more and more in favor of @MStoykov's suggestion #1. That is, instead of mutating the `SampleTags` like we currently do in `ReplaceURLWithName()`, having a method like this:
```go
func (st *SampleTags) GetCloudTags() *SampleTags {
	if st == nil || st.tags == nil {
		return nil
	}

	name, nameExist := st.tags["name"]
	url, urlExist := st.tags["url"]
	if !nameExist || !urlExist || name == url {
		return st
	}

	newTags := st.CloneTags()
	newTags["url"] = name

	return &SampleTags{tags: newTags}
}
```

It would be interesting to benchmark that in comparison to `ReplaceURLWithName()` - when both `name` and `url` tags exists, but with sub-benchmarks when they're the same or when they're different... Though even if there's some overhead, I'm still in favor of doing the copying, given how much cleaner it is.

Some other things that I notice about `SampleTags`:
- `CloneTags()` seems like it could be optimized: instead of `res := map[string]string{}`, `res:= make(map[string]string, len(st.tags))` should be slightly better, right?
- Given that we have a `Get()` method, we don't actually need to write this code in `stats/`, instead of writing the code as the method I wrote above, we can write it as a plain function in `stats/cloud`, where it's much more appropriate to live. Though, again, maybe the method would be slightly faster.

Reply:
if we actually are never going to overwrite the json buffer (which we probably shouldn't) , [sync.Once should be even faster](https://stackoverflow.com/a/53376578/5427244). But yeah it's much more important if this buffer is actually used a lot or not :). 
If we can make SampleTags immutable (good idea either way) and be able to share SampleTags between much more samples somehow it would definitely be much more beneficial but this gets into a whole refactoring so ...

Reply:
Other things that I would like to note, after looking at this some more:
1. Add tests for this :) for both samples that get aggregated and ones that don't get aggregated. They both should have their `url` equal to `name`
2. This should only happen for HttpTrail
3. Don't change HttpTrail  as well as it's also used by all other outputs, you will need to:
3.1. Just whenever using Tags run them through the GetCloudTags
3.2. Copy the HttpTrail .. .probably the easiest and less buggy way, but also probably the slowest one :( (also this should definitely be done only if actually needed)

Reply:
Suggestion: something like this in `stats/cloud/collector.go`?
```go
func UseCloudTags(source *httpext.Trail) *httpext.Trail {
	name, nameExist := source.Tags.Get("name")
	url, urlExist := source.Tags.Get("url")
	if !nameExist || !urlExist || name == url {
		return source
	}

	newTags := source.Tags.CloneTags()
	newTags["url"] = name

	dest := new(httpext.Trail)
	*dest = *source
	dest.Tags = stats.IntoSampleTags(&newTags)

	return dest
}

```

Reply:
Not certain if it matters but all the samples will have the old Tags ;) so maybe a loop over them as well ;)

Reply:
Hmm true, but it doesn't matter one jot to the cloud collector... So instead of looping over them, we can just set `dest.Samples = nil` if it worries you. 

Alternatively, we can make the `httpext.Trail.GetSamples()` method lazy, but that either means:
1. more allocations for use cases that use it multiple times (which happens often enough, with the end-of-test summary, thresholds and non-cloud outputs)
2. caching the lazily-generated slice, but then we run in the same complexities and issues like the JSON caching that necessitate synchronization

So, I'm personally for ignoring the issue or setting the `Samples` to `nil`

Reply:
I push force as this is a new implementation, please check. Also this implementation does not fix the data race problem.

Reply:
Resolved in https://github.com/loadimpact/k6/issues/1222